src/main.rs

```diff
@@ -0,0 +1,269 @@
+        Fixer {
+            // Matches lines produced by MozFormatCodeAddress().
+            re: Regex::new(r"^(.*#\d+: )(.+)\[(.+) \+0x([0-9A-Fa-f]+)\](.*)$").unwrap(),
+            file_infos: FxHashMap::default(),
+        }
+    }
+
+    /// Read the data from `file_name` and construct a `FileInfo` that we can
+    /// subsequently query. Return a description of the failing operation on
+    /// error.
+    fn build_file_info(file_name: &str) -> Result<FileInfo, String> {
+        // Get the debug session from file.
+        let msg = |op| format!("Unable to {} file {}", op, file_name);
+        let data = fs::read(file_name).map_err(|_| msg("read"))?;
+        let object = Object::parse(&data).map_err(|_| msg("parse"))?;
+        let debug_session = object
+            .debug_session()
+            .map_err(|_| msg("read debug info from"))?;
+
+        // Build the `FileInfo` from the debug session.
+        let mut interner = Interner::default();
+        let func_infos = debug_session
+            .functions()
+            .filter_map(|function| {
+                let function = function.ok()?;
+                Some(FuncInfo {
+                    address: function.address,
+                    size: function.size,
+                    mangled_name: function.name.as_str().to_string(),
+                    line_infos: function
+                        .lines
+                        .into_iter()
+                        .map(|line| LineInfo {
+                            address: line.address,
+                            line: line.line,
+                            path: interner.intern(line.file.path_str()),
+                        })
+                        .collect(),
+                })
+            })
+            .collect();
+
+        let file_info = FileInfo {
+            func_infos,
+            interner,
+        };
+
+        Ok(file_info)
+    }
+
+    // Fix stack frames within `line` as necessary. Prints any errors to stderr.
+    #[inline]
+    fn fix(&mut self, line: String) -> String {
+        // Apply the regexp.
+        let captures = if let Some(captures) = self.re.captures(&line) {
+            captures
+        } else {
+            return line;
+        };
+
+        let before = &captures[1];
+        let fn_name = &captures[2];
+        let file_name = &captures[3];
+        let address = u64::from_str_radix(&captures[4], 16).unwrap();
+        let after = &captures[5];
+
+        // If we haven't seen this file yet, parse and record its contents, for
+        // this lookup and any future lookups.
+        let file_info = match self.file_infos.entry(file_name.to_string()) {
+            Entry::Occupied(o) => o.into_mut(),
+            Entry::Vacant(v) => match Fixer::build_file_info(file_name) {
+                Ok(file_info) => v.insert(file_info),
+                Err(msg) => {
+                    eprintln!("{}", msg);
+                    return line;
+                }
+            },
+        };
```

Comment:
Split this out in a separate function with a descriptive name so you don't need the comment